fuel-vm/src/interpreter/blockchain.rs

```diff
@@ -257,161 +568,166 @@ where
+    a: Word,
+    exists: &mut Word,
+    c: Word,
+) -> Result<(), RuntimeError> {
+    let ax = checked_add_word(a, Bytes32::LEN as Word)?;
 
-        state_read_qword(&contract_id, storage, memory, &mut registers[rb], input)?;
-        self.inc_pc()
+    if ax > VM_MAX_RAM {
+        return Err(PanicReason::MemoryOverflow.into());
     }
 
-    pub(crate) fn state_write_word(&mut self, a: Word, rb: RegisterId, c: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(rb)?;
-
-        let ax = checked_add_word(a, Bytes32::LEN as Word)?;
-
-        if ax > VM_MAX_RAM {
-            return Err(PanicReason::MemoryOverflow.into());
-        }
-
-        let (a, ax) = (a as usize, ax as usize);
-        let (d, dx) = self.internal_contract_bounds()?;
+    let (a, ax) = (a as usize, ax as usize);
+    let (d, dx) = internal_contract_bounds(context, fp)?;
 
-        // Safety: Memory bounds logically verified by the interpreter
-        let contract = unsafe { ContractId::as_ref_unchecked(&self.memory[d..dx]) };
-        let key = unsafe { Bytes32::as_ref_unchecked(&self.memory[a..ax]) };
+    // Safety: Memory bounds logically verified by the interpreter
+    let contract = unsafe { ContractId::as_ref_unchecked(&memory[d..dx]) };
+    let key = unsafe { Bytes32::as_ref_unchecked(&memory[a..ax]) };
 
-        let mut value = Bytes32::default();
+    let mut value = Bytes32::default();
 
-        value[..WORD_SIZE].copy_from_slice(&c.to_be_bytes());
-
-        let result = self
-            .storage
-            .merkle_contract_state_insert(contract, key, &value)
-            .map_err(RuntimeError::from_io)?;
-
-        self.registers[rb] = result.is_some() as Word;
-
-        self.inc_pc()
-    }
-
-    pub(crate) fn state_write_qword(&mut self, a: Word, rb: RegisterId, c: Word, d: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(rb)?;
-        let contract_id = *self.internal_contract()?;
-        let input = StateWriteQWord::new(a, c, d)?;
-        let Self {
-            ref mut storage,
-            ref mut memory,
-            ref mut registers,
-            ..
-        } = self;
+    value[..WORD_SIZE].copy_from_slice(&c.to_be_bytes());
 
-        state_write_qword(&contract_id, storage, memory.as_mut(), &mut registers[rb], input)?;
-        self.inc_pc()
-    }
+    let result = storage
+        .merkle_contract_state_insert(contract, key, &value)
+        .map_err(RuntimeError::from_io)?;
 
-    pub(crate) fn timestamp(&mut self, ra: RegisterId, b: Word) -> Result<(), RuntimeError> {
-        Self::is_register_writable(ra)?;
+    *exists = result.is_some() as Word;
 
-        self.block_height()
-            .and_then(|c| (b <= c as Word).then_some(()).ok_or(PanicReason::TransactionValidity))?;
+    inc_pc(pc)
+}
 
-        let b = u32::try_from(b).map_err(|_| PanicReason::ArithmeticOverflow)?;
+pub(crate) fn timestamp(
+    storage: &impl InterpreterStorage,
+    block_height: u32,
```

Comment:
While passing block_height instead of fetching it from storage here does make unit testing of b > block_height easier, it seems somewhat strange to me to have several layers of functions plucking one thing out of storage at a time before calling the next function. 

Personally I'd prefer to keep the primary domain logic contained within in these selfless functions as much as possible, and limit the logic in the &self methods to the bare minimum needed to pass references to registers, memory & storage from the Interpreter instance. The more fragmented these functions become, the harder it is to analyze and also potentially leave more room for error in future refactors if anything can be passed in for block height.